lib/ASTGen/Sources/ASTGen/Attrs.swift

```diff
@@ -0,0 +1,1132 @@
+    guard
+      var args = node.arguments?.as(LabeledExprListSyntax.self)?[...],
+      let arg = args.popFirst()
+    else {
+      // TODO: Diagnose.
+      return nil
+    }
+
+    let isRaw: Bool
+    if let label = arg.label {
+      if label.rawText == "raw" {
+        isRaw = true
+      } else {
+        // TODO: Diagnose.
+        return nil
+      }
+    } else {
+      isRaw = false
+    }
+    guard let name = self.generateStringLiteralTextIfNotInterpolated(expr: arg.expression) else {
+      // TODO: Diagnose.
+      return nil
+    }
+    guard args.isEmpty else {
+      // TODO: Diagnose
+      return nil
+    }
+
+    return .createParsed(
+      self.ctx,
+      atLoc: self.generateSourceLoc(node.atSign),
+      range: self.generateSourceRange(node),
+      name: name,
+      isRaw: isRaw
+    )
+  }
+
+  func generateSPIAccessControlAttr(attribute node: AttributeSyntax) -> BridgedSPIAccessControlAttr? {
+    let spiName: BridgedIdentifier? = self.generateSingleAttrOption(attribute: node) {
+      self.generateIdentifier($0)
+    }
+    guard let spiName else {
+      return nil
+    }
+    return .createParsed(
+      self.ctx,
+      atLoc: self.generateSourceLoc(node.atSign),
+      range: self.generateSourceRange(node),
+      spiGroupName: spiName
+    )
+  }
+
+  func generateSwiftNativeObjCRuntimeBaseAttr(attribute node: AttributeSyntax) -> BridgedSwiftNativeObjCRuntimeBaseAttr? {
+    let name: BridgedIdentifier? = self.generateSingleAttrOption(attribute: node) {
+      self.generateIdentifier($0)
+    }
+    guard let name else {
+      return nil
+    }
+    return .createParsed(
+      self.ctx,
+      atLoc: self.generateSourceLoc(node.atSign),
+      range: self.generateSourceRange(node),
+      name: name
+    )
+  }
+
+  func generateSimpleDeclAttr(attribute node: AttributeSyntax, kind: BridgedDeclAttrKind) -> BridgedDeclAttribute? {
+    // TODO: Diagnose extraneous arguments.
+    // TODO: Diagnose if `kind` is a modifier.
+    return BridgedDeclAttribute.createSimple(
+      self.ctx,
+      kind: kind,
+      atLoc: self.generateSourceLoc(node.atSign),
+      nameLoc: self.generateSourceLoc(node.attributeName)
+    )
+  }
+
+  func generateCustomAttr(attribute node: AttributeSyntax) -> BridgedCustomAttr? {
+    guard
+      var args = node.arguments?.as(LabeledExprListSyntax.self)?[...]
+    else {
+      // TODO: Diagnose.
+      return nil
+    }
+    fatalError("unimplemented")
+  }
+
+  func generateStringLiteralTextIfNotInterpolated(expr node: some ExprSyntaxProtocol) -> BridgedStringRef? {
+    if let segments = node.as(SimpleStringLiteralExprSyntax.self)?.segments {
+      return extractRawText(segments).bridged
+    } else if let segments = node.as(StringLiteralExprSyntax.self)?.segments,
+      segments.all({ $0.is(StringSegmentSyntax.self) })
+    {
+      return extractRawText(segments).bridged
+    }
+    return nil
+  }
+
+  func generateConsumingAttrOption<R>(
```

Comment:
I think a doc comment for these `generateConsuming` functions would help. Especially to clarify whether they care about ordering of arguments or not.